tensorboard/loader.py

```diff
@@ -0,0 +1,414 @@
+    """Returns byte length of file.
+
+    This is guaranteed to return a number greater than or equal to the
+    offset of the last record returned by get_next_record().
+
+    This method can be called after the instance has been closed.
+
+    Raises:
+      IOError: If file has shrunk from last offset or start offset, or
+          this instance is closed.
+
+    :rtype: int
+    """
+    length = tf.gfile.Stat(self.path).length
+    if length < self.get_offset():
+      raise IOError('File shrunk %d < %d: %s' %
+                    (length, self._offset, self.path))
+    return length
+
+  def close(self):
+    """Closes record reader if open.
+
+    Further reads are not permitted after this method is called.
+    """
+    if self._is_closed:
+      return
+    if self._reader is not None:
+      self._reader.Close()
+    self._is_closed = True
+    self._reader = None
+
+  def _open(self):
+    with tf.errors.raise_exception_on_not_ok_status() as status:
+      return tf.pywrap_tensorflow.PyRecordReader_New(
+          tf.resource_loader.readahead_file_path(tf.compat.as_bytes(self.path)),
+          self._offset, '', status)
+
+  def __str__(self):
+    return u'RecordReader{%s}' % self.path
+
+
+@util.closeable
+@six.python_2_unicode_compatible
+class BufferedRecordReader(object):
+  """Wrapper around RecordReader that does threaded read-ahead.
+
+  This class implements the same interface as RecordReader. It prevents
+  remote file systems from devastating loader performance. It does not
+  degrade throughput on local file systems.
+
+  The thread is spawned when the first read operation happens. The
+  thread will diligently try to buffer records in the background. Its
+  goal is to sleep as much as possible without blocking read operations.
+
+  This class is thread safe. It can be used from multiple threads
+  without any need for external synchronization.
+  """
+
+  READ_AHEAD_AGGRESSION = 2.3
+  READ_AHEAD_BYTES = 16 * 1024 * 1024
+  STAT_INTERVAL_SECONDS = 4.0
+
+  def __init__(self, path,
+               start_offset=0,
+               read_ahead=None,
+               clock=time.time):
+    """Creates new instance.
+
+    The i/o thread is not started until the first read happens.
+
+    Args:
+      path: Path of file. This can be on a remote file system if the
+          TensorFlow build supports it.
+      start_offset: Byte offset to seek in file once it's opened.
+      read_ahead: The number of record bytes to buffer into memory
+          before the thread starts blocking. This value must be >0 and
+          the default is BufferedRecordReader.DEFAULT_READ_AHEAD.
+      clock: Function returning a float with the number of seconds
+          since the UNIX epoch in zulu time.
+
+    :type path: str
+    :type start_offset: int
+    :type read_ahead: int
+    :type clock: () -> float
+    """
+    self.path = tf.compat.as_text(path)
+    self._read_ahead = read_ahead or BufferedRecordReader.READ_AHEAD_BYTES
```

Comment:
You can't do that because default values for arguments are values, not expressions, and so are evaluated at function definition time; due to the way that class desugaring works, `BufferedRecordReader` is not in scope at function definition time. Try:
```py
>>> class Foo(object):
...     BAR = 6
...     def quux(self, baz=Foo.BAR):
...         return baz + 1
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 3, in Foo
NameError: name 'Foo' is not defined
```